File: app/src/main/java/star/iota/acgrip/ui/fan/FanPresenter.java
package star.iota.acgrip.ui.fan;

import android.os.AsyncTask;

import java.util.List;

import star.iota.acgrip.service.DataService;

class FanPresenter implements FanContract.Presenter {

    private final FanContract.View view;
    private AsyncTask<Void, Void, List<FansBean>> asyncTask;

    FanPresenter(FanContract.View view) {
        this.view = view;
    }


    @Override
    public void request(final String url) {
        asyncTask = new AsyncTask<Void, Void, List<FansBean>>() {
            @Override
            protected List<FansBean> doInBackground(Void... params) {
                DataService source = new DataService();
                return source.getFans(url);
            }

            @Override
            protected void onPostExecute(List<FansBean> items) {
                if (items == null) {
                    view.error();
                } else if (items.size() == 0) {
                    view.noData();
                } else {
                    view.success(items);
                }
            }
        };
        asyncTask.execute();
    }

    @Override
    public void cancel() {
        if (asyncTask != null && !asyncTask.isCancelled()) {
            asyncTask.cancel(true);
        }
    }
}

File: app/src/main/java/star/iota/acgrip/ui/fan/FanPresenter.kt
package star.iota.acgrip.ui.fan

import io.reactivex.Flowable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import star.iota.acgrip.service.DataRepository

internal class FanPresenter(private val view: FanContract.View) : FanContract.Presenter {

    companion object {
        private val compositeDisposable = CompositeDisposable()
    }

    override fun unsubscribe() {
        Companion.compositeDisposable.clear()
    }

    override fun request(url: String) {
        Companion.compositeDisposable.add(
                Flowable.just(DataRepository())
                        .map { service -> service.getFans(url) }
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(
                                { items ->
                                    if (items != null) {
                                        if (items.isEmpty()) view.noData()
                                        else view.success(items)
                                    } else {
                                        view.error("未知错误")
                                    }
                                },
                                { error ->
                                    view.error(error.message)
                                }
                        )
        )
    }
}
